libs/sdk-ui-kit/src/@ui/UiFocusTrap/UiFocusTrap.tsx

```diff
@@ -0,0 +1,125 @@
+
+    useEffect(() => {
+        const getNextElement = (focusableElements, firstElement, lastElement, shiftKey: boolean) => {
+            const elements = Array.from(focusableElements);
+            const currentIndex = elements.indexOf(document.activeElement as HTMLElement);
+
+            let nextElement: HTMLElement | undefined;
+            if (shiftKey) {
+                // Shift + Tab - moving backwards
+                nextElement = currentIndex <= 0 ? lastElement : elements[currentIndex - 1];
+            } else {
+                // Tab - moving forwards
+                nextElement =
+                    currentIndex === elements.length - 1 ? firstElement : elements[currentIndex + 1];
+            }
+            return nextElement;
+        };
+
+        const handleKeyDown = (event: KeyboardEvent) => {
+            // get fresh focusable elements set as they could change in meantime
+            const { firstElement, lastElement, focusableElements } = getFocusableElements(trapRef.current);
+            if (event.key === "Tab") {
+                const nextElement = getNextElement(
+                    focusableElements,
+                    firstElement,
+                    lastElement,
+                    event.shiftKey,
+                );
+
+                if (nextElement) {
+                    event.preventDefault();
+                    nextElement.focus();
+                    // handle elements which are focusable but not reachable because eg. out of view port or disabled
+                    // It is eg. virtualized list inside the trap which renders some invisible items which are in focusableElements collection but when focused they do nothing -> skip them and reach some really focusable element
+                    let newNextElement = nextElement;
+                    let attempts = 0;
+                    const maxAttempts = focusableElements.length;
+                    while (newNextElement !== document.activeElement && attempts < maxAttempts) {
+                        attempts++;
+                        const currentIndex = Array.from(focusableElements).indexOf(
+                            newNextElement as HTMLElement,
+                        );
+                        const nextIndex = event.shiftKey
+                            ? (currentIndex - 1 + focusableElements.length) % focusableElements.length
+                            : (currentIndex + 1) % focusableElements.length;
+                        newNextElement = focusableElements[nextIndex];
+                        newNextElement?.focus();
+                    }
+                }
+            } else if (event.key === "Escape") {
+                if (onDeactivate) {
+                    onDeactivate();
+                }
+                if (returnFocusTo.current) {
+                    returnFocusTo.current.focus();
+                }
+            }
+        };
+
+        document.addEventListener("keydown", handleKeyDown);
```

Comment:
Shouldn't `handleKeyDown` rather be wrapped in `useCallback`?

Reply:
react vs real DOM events issue. My experience is, that order or synthetic ones can be sometimes tricky